oak/client/sgx_application_client.cc

```diff
@@ -0,0 +1,167 @@
+#include "asylo/identity/enclave_assertion_authority_configs.h"
+#include "asylo/identity/init.h"
+#include "asylo/identity/sgx/sgx_identity_util.h"
+#include "asylo/identity/util/sha256_hash.pb.h"
+#include "asylo/util/logging.h"
+#include "oak/client/authorization_bearer_token_metadata.h"
+#include "oak/client/policy_metadata.h"
+#include "oak/common/nonce_generator.h"
+#include "oak/common/policy.h"
+
+constexpr size_t kPerChannelNonceSizeBytes = 32;
+// Debug MRSIGNER value derived from the SGX test key (will change after changing SVN and PRODID).
+const char* kDebugMrSigner = "83d719e77deaca1470f6baf62a4d774303c899db69020f9c70ee1dfc08c7ce9e";
+
+// TODO: Use same function from asylo, when it will become public.
+bool Sha256HashFromHexString(const std::string& hex, asylo::Sha256HashProto* h) {
+  if (hex.size() != 64) {
+    return false;
+  }
+  for (auto ch : hex) {
+    if (std::isxdigit(ch) == 0) {
+      return false;
+    }
+  }
+  h->set_hash(absl::HexStringToBytes(hex).c_str(), hex.size() / 2);
+  return true;
+}
+
+namespace oak {
+
+SgxApplicationClient::SgxApplicationClient(
+    /*intel_public_key,*/
+    std::vector<std::string> mrenclave_strings) {
+  // Initialize assertion authorities
+  this->InitializeAssertionAuthorities();
+
+  // Initialize credentials
+  auto channel_credentials = this->CreateChannelCredentials(mrenclave_strings);
+  auto call_credentials = this->CreateCallCredentials();
+  this->credentials_ = grpc::CompositeChannelCredentials(channel_credentials, call_credentials);
+}
+
+std::shared_ptr<grpc::Channel> SgxApplicationClient::CreateChannel(std::string address) {
+  return grpc::CreateChannel(address, this->credentials_);
+}
+
+asylo::EnclaveAssertionAuthorityConfig SgxApplicationClient::GetNullAssertionAuthorityConfig() {
+  asylo::EnclaveAssertionAuthorityConfig test_config;
+  asylo::SetNullAssertionDescription(test_config.mutable_description());
+  return test_config;
+}
+
+void SgxApplicationClient::InitializeAssertionAuthorities() {
+  LOG(INFO) << "Initializing assertion authorities";
+  std::vector<asylo::EnclaveAssertionAuthorityConfig> configs = {
+      GetNullAssertionAuthorityConfig(),
+      // TODO: Provide SGX remote Assertion Authorities when available.
+      // GetSgxRemoteAssertionAuthorityConfig()
+  };
+
+  auto status = asylo::InitializeEnclaveAssertionAuthorities(configs.begin(), configs.end());
+  if (!status.ok()) {
+    LOG(QFATAL) << "Could not initialize assertion authorities";
+  }
+  LOG(INFO) << "Assertion authorities initialized";
+}
+
+// TODO: Add CPUSVN as a parameter.
+asylo::EnclaveIdentityExpectation SgxApplicationClient::CreateSgxIdentityExpectation(
+    std::string& mrenclave_string) const {
+  asylo::SgxIdentity sgx_identity;
+
+  // Assign code identity.
+  auto code_identity = sgx_identity.mutable_code_identity();
+  if (!Sha256HashFromHexString(mrenclave_string, code_identity->mutable_mrenclave())) {
+    LOG(QFATAL) << "Bad MRENCLAVE";
```

Comment:
Maybe return `StatusOr<asylo::EnclaveIdentityExpectation>` from the function instead of this failure?